### PR TITLE
feat: added inmemory cache and also added add wallet queue worker.

### PR DESCRIPTION
**Pull request by shubhampatwa**

### CODE REVIEW

**Comment by spunkypanda** [comment]
In this PR, the wallet addresses are being stored as an array. I think we should store wallet addresses as a dictionary with the address as key and its status as its value. The default value of the status can be 1 or true.

With a dictionary, the lookup times for a addresses in the cache is O(1). With an array, its O(n).

This status should also be stored in the database, with default value as 1 or true.

We also need a method to check if an address record exists in inMemoryCache[PACMAN_WALLETS] and its status is 1 or true.

File: src/utils/in-memory-cache.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  if (walletExists) {
+    walletExists.push(address);
+    cache.set(PACMAN_WALLETS, walletExists);
+  } else {
+    cache.set(PACMAN_WALLETS, [address]);
+  }
+};
+
+export const getCacheWallets = async (): Promise<string[]> => {
+  let wallets: string[] = cache.get(PACMAN_WALLETS);
```
Comment (spunkypanda) [inline src/utils/in-memory-cache.ts]: We should store cachedWallets as a dictionary due to O(1) lookup time

```suggestion
  let wallets: Record<string, boolean> = cache.get(PACMAN_WALLETS);
```

File: src/utils/in-memory-cache.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  };
+};
+
+export const cache = inMemoryCache();
+
+export const updateWalletCache = async (address: string) => {
+  await saveWallet(address);
+  const walletExists: string[] = cache.get(PACMAN_WALLETS);
+  if (walletExists) {
+    walletExists.push(address);
```
Comment (spunkypanda) [inline src/utils/in-memory-cache.ts]:
```suggestion
    walletExists[address] = true;
```

File: src/utils/in-memory-cache.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    cache.set(PACMAN_WALLETS, walletExists);
+  } else {
+    cache.set(PACMAN_WALLETS, [address]);
```
Comment (spunkypanda) [inline src/utils/in-memory-cache.ts]:
```suggestion
    cache.set(PACMAN_WALLETS, { [address]: true });
```

File: src/utils/in-memory-cache.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      cache.set(key, val);
+    },
+  };
+};
+
+export const cache = inMemoryCache();
+
+export const updateWalletCache = async (address: string) => {
+  await saveWallet(address);
+  const walletExists: string[] = cache.get(PACMAN_WALLETS);
```
Comment (spunkypanda) [inline src/utils/in-memory-cache.ts]:
```suggestion
  const walletExists: Record<string, boolean> = cache.get(PACMAN_WALLETS);
```

File: src/migrations/1678970045740_wallet.sql [hunk 5]
```diff
@@ -0,0 +1,11 @@
+-- Up Migration
+CREATE TABLE "pacman_wallets" (
+  "address" VARCHAR NOT NULL,
+  "created_at" timestamp with time zone DEFAULT NOW()
```
Comment (spunkypanda) [inline src/migrations/1678970045740_wallet.sql:5]:
```suggestion
  "status" smallint DEFAULT 1,
  "created_at" timestamp with time zone DEFAULT NOW()
```

**Review by spunkypanda — CHANGES_REQUESTED** [review]
Lets store the cached wallets as a dictionary instead of an array. I've added some code suggestions to help you out

File: src/utils/in-memory-cache.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+};
+
+export const getCacheWallets = async (): Promise<string[]> => {
+  let wallets: string[] = cache.get(PACMAN_WALLETS);
+  if (!wallets) {
+    wallets = await idb.many("select address from pacman_wallets");
+  }
+  return wallets ?? [];
+};
+
```
Comment (spunkypanda) [inline src/utils/in-memory-cache.ts:42]: We need to have a method to quickly check whether a wallet address is a tracked wallet

```suggestion
export const isCachedWallet = async (address: string) => {
  const cachedWallets: Record<string, boolean> = cache.get(PACMAN_WALLETS);
  
  if (!cachedWallets || !cachedWallets[address]) return false;
  
  return true;
};
```

File: src/utils/in-memory-cache.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    cache.set(PACMAN_WALLETS, [address]);
+  }
+};
+
+export const getCacheWallets = async (): Promise<string[]> => {
+  let wallets: string[] = cache.get(PACMAN_WALLETS);
+  if (!wallets) {
+    wallets = await idb.many("select address from pacman_wallets");
+  }
+  return wallets ?? [];
```
Comment (spunkypanda) [inline src/utils/in-memory-cache.ts]:
```suggestion
  return wallets ? Object.keys(wallets) : [];
```

File: src/utils/in-memory-cache.ts [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    cache.set(PACMAN_WALLETS, walletExists);
+  } else {
+    cache.set(PACMAN_WALLETS, [address]);
+  }
+};
+
+export const getCacheWallets = async (): Promise<string[]> => {
+  let wallets: string[] = cache.get(PACMAN_WALLETS);
+  if (!wallets) {
+    wallets = await idb.many("select address from pacman_wallets");
```
Comment (spunkypanda) [inline src/utils/in-memory-cache.ts]:
```suggestion
    const pacmanWallets = await idb.many("select address, status from pacman_wallets");
    
    wallets = pacmanWallets.reduce((acc, wallet) => {
      acc[wallet.address] = !!wallet.status;
      return acc;
    }, {});
```

File: src/utils/in-memory-cache.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+      acc[address] = true;
+      return acc;
+    }, {}) ?? {}
+  );
+};
+
+export const saveWallet = async (address: string) => {
+  return idb.none(
+    `
+      INSERT INTO pacman_wallets (
```
Comment (ptsayli) [inline src/utils/in-memory-cache.ts]: Can we avoid adding pacman related information on the indexer? @shubhampatwa

**Review by ptsayli — CHANGES_REQUESTED** [review]

File: src/utils/in-memory-cache.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    cache.set(PACMAN_WALLETS, walletExists);
+  } else {
+    cache.set(PACMAN_WALLETS, { [address]: true });
+  }
+};
+
+export const getCacheWallets = async (): Promise<Record<string, boolean>> => {
+  let wallets: string[] = cache.get(PACMAN_WALLETS);
+  if (!wallets) {
+    wallets = await idb.many("select address from pacman_wallets where status = 1");
```
Comment (spunkypanda) [inline src/utils/in-memory-cache.ts]: Ran this locally and its throwing an error when there are no records in this table. We will need to replace the method here

```suggestion
    wallets = await idb.manyOrNone("select address from pacman_wallets where status = 1");
```

File: src/utils/in-memory-cache.ts (hunk 9 above)
Comment (spunkypanda) [reply src/utils/in-memory-cache.ts]: @ptsayli @shubhampatwa I've pushed a commit to this branch changing name of the table to `tracked_wallets`. This commit also fixes a minor bug'

File: src/utils/in-memory-cache.ts (hunk 10 above)
Comment (spunkypanda) [reply src/utils/in-memory-cache.ts]: @shubhampatwa Pushed this change in a commit here as I needed this in my branch

**Review by spunkypanda — APPROVED** [review]

**Review by ptsayli — APPROVED** [review]

**Pull request merged**